src/liballoc/boxed.rs

```diff
@@ -428,7 +428,15 @@ impl<T: ?Sized> Box<T> {
     #[stable(feature = "box_raw", since = "1.4.0")]
     #[inline]
     pub fn into_raw(b: Box<T>) -> *mut T {
-        Box::into_raw_non_null(b).as_ptr()
+        let b = mem::ManuallyDrop::new(b);
+        let mut unique = b.0;
+        // Box is kind-of a library type, but recognized as a "unique pointer" by
+        // Stacked Borrows.  This function here corresponds to "reborrowing to
+        // a raw pointer", but there is no actual reborrow here -- so
+        // without some care, the pointer we are returning here still carries
+        // the tag of `b`, with `Unique` permission.
+        // We round-trip through a mutable reference to avoid that.
+        unsafe { unique.as_mut() as *mut T }
```

Comment:
@RalfJung I moved this from `into_unique`. Is it still needed for `into_raw`? Does this look correct?

Reply:
Yes this is still needed I think, we still want an explicit ref-to-raw cast to actually "mark" this as a raw pointer for the aliasing model. It looks correct.

Reply:
In the light it might be slightly clearer to implement `leak` directly, similar to this method, and then implement `into_raw` as `Box::leak(b) as *mut T`?

Reply:
Good point, except that `unique.as_mut()` returns a lifetime bound to `unique` on the stack and `leak()` needs an unbound lifetime. Would `&mut *unique.as_ptr()` satisfy Stacked Borrows?

Reply:
Yes that should work, as a reference is explicitly created.

The issue here is that `Box`, for stacked borrows, is almost like a reference, but for rustc when accessing its `.0` fields it's more like a raw pointer -- so if we are not careful we end up with a mutable reference effectively transmuted to a raw pointer, which still expects to have unique access. An explicit `&mut` here (or the one in `Unique::as_mut()`) solves that by making things "coherent" again -- now both rustc and Stacked Borrows agree this is a mutable reference.